concepts/functions/about.md

```diff
@@ -1,3 +1,93 @@
 # About
 
-TODO
+Functions are bodies of code containing one or more [statements or expressions][statements or expressions]. A function may optionally
+[return an value][return value].
+
+Rust style is to use [snake case][snake case] for a function name, which is prefaced by the `fn` keyword for a function definition. For example
+
+```rust
+fn do_nothing() {}
+```
+
+The parentheses surround zero or more parameters, separated by commas. A parameter is a binding to a value of a particular type which is passed into the function.
+Even if there are no parameters the parentheses are still required in the function definition. The actual value passed to a parameter is called
+an argument.
+
+The combination of parameters and return value is known as the function's signature. A function's signature requires that each parameter must
+have its type annotated.
+
+The curly braces enclose the body of the function: its statements and expressions. Even if the function contains no statements or expressions the
+curly braces are still required for the function definition.
+
+In the following example the function takes in one parameter of type `i32`, binds it to the name `value`, and prints it.
+
+```rust
+fn print_integer(value: i32) {
+    println!("{:?}", value);
+}
+```
+
+Note the parameter's definition. Each parameter is defined in the format `name: Type`.
+
+A function can also return a value. By default, the output of the final expression is returned.
+In the following example the function has one `i32` parameter and returns its double.
+
+```rust
+fn double_integer(value: i32) -> i32 {
+    value * 2
+}
+```
+
+The `-> i32` indicates that the function returns an `i32`. Unlike parameters, the returned value is not named.
+
+It is possible to exit from a function early with the `return` keyword, like so:
+
+```rust
+fn long_function() -> i32 {
+    let some_condition = false;
+    // code snipped
+    if (some_condition) {
+        return 0;
+    }
+    // more code snipped
+    42
+}
+```
+
+`const fn` is used to define a [constant function][constant function], which can be evaluated at compile time. Maing a function constant
+is a way of declaring that the function won't change in ways that are invalid for a `const fn`. Since this may prevent future optimizations,
+making a function `const` should be considered with care. Going back to `double_integer`, by making it a `const fn` we can now use it to set a
```

Comment:
> Since this may prevent future optimizations, making a function `const` should be considered with care.

I'm not sure this is accurate. Do you have an example of an optimization prevented by making a function `const`?

Reply:
Got it from here a reddit discussion. Personally, I haven't had much use for `const fn`, so I'm somewhat at a loss to satisfy your request to add a note explaining what that can be good for. You'd be much better qualified to explain that than I.

https://www.reddit.com/r/rust/comments/iksmgk/psa_what_is_const_fn/#:~:text=const%20fn%20is%20a%20contract,think%20carefully%20before%20using%20it.

Reply:
The point being made in that PR is that a `const` annotation is part of semantic versioning. If you later refactor a public `const fn` in a way which makes it impossible to implement as `const`, you're going to need to bump your major version to remove that annotation. I don't think that's a relevant concern here. AFAIK we don't even teach semantic versioning in this track.

Reply:
I'm happy to get rid of all of that. Do you have something to put in for what a `const fn` is good for?

Reply:
```rust
const fn compute_data_checksum() -> u128 {
  const DATA: &[u8] = include_bytes!("my_big_data_file");
  // checksum implementation is left as an exercise for the reader
}

/// This checksum is used to validate that the user has not tampered with proprietary configuration.
pub const DATA_CHECKSUM: u128 = compute_data_checksum();
```

Reply:
Note that in that example, the only data that will actually appear in the produced binary is the checksum, not the big data file which produced it.